endpoints/openrtb2/auction.go

```diff
@@ -434,6 +440,53 @@ func validateSChains(req *openrtb_ext.ExtRequest) error {
 	return err
 }
 
+// validateCustomRates discards invalid 3-digit currency codes found in the bidRequest.ext.prebid.currency
+// field. If all of them were discarded and bidRequest.ext.prebid.currency was set to false, it means
+// we have no currencies to work with and this function returns a fatal BadInput error
+func validateCustomRates(bidReqCurrencyRates *openrtb_ext.ExtRequestCurrency) error {
+
+	if bidReqCurrencyRates == nil {
+		return nil
+	}
+
+	var badCurrencyCodes []string
+
+	for fromCurrency, rates := range bidReqCurrencyRates.ConversionRates {
+
+		// Check if fromCurrency is a valid 3-letter currency code
+		_, err := currency.ParseISO(fromCurrency)
+		if err != nil {
+			badCurrencyCodes = append(badCurrencyCodes, fromCurrency)
+			delete(bidReqCurrencyRates.ConversionRates, fromCurrency)
+		}
+
+		// Check if currencies mapped to fromCurrency are valid 3-letter currency codes
+		for toCurrency := range rates {
+			_, err := currency.ParseISO(toCurrency)
+			if err != nil {
+				badCurrencyCodes = append(badCurrencyCodes, toCurrency)
+				delete(rates, toCurrency)
+			}
+		}
+
+		// If all of fromCurrency's mapped currencies were invalid, remove fromCurrency map entry
+		if len(rates) == 0 {
+			delete(bidReqCurrencyRates.ConversionRates, fromCurrency)
+		}
+	}
+
+	// If we are retricted to only use custom currency rates but the currency rate map is empty, return error
+	customRatesOnly := bidReqCurrencyRates.UsePBSRates != nil && !(*bidReqCurrencyRates.UsePBSRates)
+	if customRatesOnly && len(bidReqCurrencyRates.ConversionRates) == 0 {
+		if len(badCurrencyCodes) > 0 {
+			return &errortypes.BadInput{Message: fmt.Sprintf("Required custom currency rates are all invalid. %s", strings.Join(badCurrencyCodes, ","))}
```

Comment:
But to reach this branch `bidReqCurrencyRates.ConversionRates` must be empty which means either it was empty from the beginning, which leads to returning the error in line 484, or it got emptied after finding all of the 3-letter currency codes to be wrong that resolves in returning error in 482. Do you agree with this approach? We could still reword the error message or even refactor this part if needed.

Reply:
I don't think that's necessarily true. As the code is currently, a `ConversionRates` entry will be removed if the base currency is wrong even if the mapped currencies are correct. For example:
```
ConversionRates: map[string]map[string]float64{
    "FOO": {
        "GBP": 1.2,
        "MXN": 0.05,
        "JPY": 0.95,
    },
}
```
If I'm reading the code correctly, this will cause the error message to be reached when `UsePBSRates` is true. While the base currency of FOO is invalid, the mapped currencies of GBP, MXN, and JPY are valid.

Reply:
I don't think that's necessarily true. As the code is currently, a `ConversionRates` entry will be removed if the base currency is wrong even if the mapped currencies are correct. For example:
```
ConversionRates: map[string]map[string]float64{
    "FOO": {
        "GBP": 1.2,
        "MXN": 0.05,
        "JPY": 0.95,
    },
}
```
If I'm reading the code correctly, this will cause the error message to be reached when `UsePBSRates` is true. While the base currency of FOO is invalid, the mapped currencies of GBP, MXN, and JPY are valid.

Reply:
This part of the code has been removed because we are now immediately returning a bad input error when the first non-valid code is found.